oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,344 @@
+| Status          | Draft                                                  |
++-----------------+--------------------------------------------------------+
+| Type            | Best Practice                                          |
++-----------------+--------------------------------------------------------+
+| Created         | 2018-03-27                                             |
++-----------------+--------------------------------------------------------+
+| `Review Period` | 2018-03-27 - 2018-04-20                                |
++-----------------+--------------------------------------------------------+
+| `Resolution`    |                                                        |
++-----------------+--------------------------------------------------------+
+
+Abstract
+========
+
+Proposes best practices for declaring and maintaining dependencies on other
+Python packages in Open edX software repositories.
+
+Motivation
+==========
+
+The Open edX project includes dozens of Python software repositories, most of
+which depend on certain other Python packages being installed in order to
+function correctly.  The simple methods we originally used to do this have
+assorted drawbacks that have repeatedly caused problems over the past few
+years: accidental upgrades to incompatible versions, strict installation
+requirements that restrict the ability of downstream packages to manage their
+own dependency versions, lack of clarity regarding the full set of packages
+actually depended upon, etc.
+
+Outlined here is a recommended standard for declaring dependencies on other
+Python packages which resolves most of these issues and will let us make all
+the Open edX Python packages consistent with each other (and many other open
+source Python projects) for ease of understanding and maintenance.
+
+Specification
+=============
+
+The key to successful Python dependency management is to break it down into
+four parts:
+
+1. Identify the different contexts in which dependencies will need to be
+   installed.
+2. For each of these contexts, declare the direct dependencies that will be
+   needed.  Use the least restrictive constraints which should allow pip to
+   install a working set of dependencies.
+3. Auto-generate from the high-level dependencies declaration a complete
+   set of exact package versions that are known to work for a Python
+   `virtualenv`_ created for that context.
+4. Automate updates of the detailed dependencies listing for each context.
+
+.. _virtualenv: https://virtualenv.pypa.io/
+
+Identify Usage Contexts
+-----------------------
+
+The dependencies of Python software are typically installed or run in a
+variety of different contexts over the course of developing and using it.
+The set of dependencies needed to perform a task can easily vary between these
+contexts.  For example:
+
+* Just the standard set of core dependencies for execution on a production
+  server to perform its primary purpose (``base.in``)
```

Comment:
You have not explained what "base.in" is, or why you are mentioning it here.  Maybe a paragraph at the top giving the overview ("We'll use pip-tools to manage multiple contexts."), then you can tie these filenames into that.

Reply:
Yeah, I was hoping to avoid front-loading too much description of the overall process, but there probably does need to be some advance mention of what these names are for.